book/src/solana-scripts.md

```diff
@@ -0,0 +1,108 @@
+# Solana Scripts
+
+This architecture describes how the Pipeline runtime supports scripts. 
+
+In Solana programs have absolute control over the state assigned to them.  The
+goal of this architecture is provide a simple way for on-chain programs to
+interact with other programs without violating the state ownership guarantees.
+
+Contract composition has led to multiple multi-million dollar bugs in Ethereum
+
+* https://www.parity.io/security-alert-2/
+
+## Scripts
+
+The purpose of scripts is to execute logic between instructions.
+
+For example, assignment of tokens to the winner based on an output of a game can
+be defined as a script.
+
+### `Loader::CreateScript`
+
+`Loader::CreateScript` designates that the loaded executable bytecode is a
+script. The difference between scripts and programs is that script execution
```

Comment:
I don't understand why we wouldn't extend `LoaderInstruction::Finalize` with first-level program dependencies. That is, `LoaderInstruction::Finalize` -> `LoaderInstruction::Finalize(Vec<Pubkey>)`.  That would tell the loader to load all the executable programs that might be invoked by this newly finalized program.

Reply:
Just linking programs together doesn’t really help. 

In the Move paper they talk about the verification step that checks that the user authorized the call.  We need to convey that as well.